### PR TITLE
handle node deletion

### LINKED ISSUE #1152: Unable to delete seccompprofile if one of node is deleted
#### What happened:

Unable to delete the seccompprofiles if one of the node deleted/updated because of the finalizers is not updated with the latest nodes. I'm not sure if this bug or as per design.
Thank you

#### What you expected to happen:

Possible if SPO update seccompprofiles on the Finalizers part to the latest k8s nodes?

#### How to reproduce it (as minimally and precisely as possible):

- Start minikube with 3 nodes
```$ minikube  start --kubernetes-version=v1.21.14 --nodes=3 --memory 4000 --cpus```
- Install SPO
- Check the default SeccomProfiles, there are 3 finalizers
```
$ kubectl get seccompprofiles -n security-profiles-operator log-enricher-trace -o json | jq .metadata.finalizers
[
            "minikube-m03-delete",
            "minikube-m02-delete",
            "minikube-delete"
]
```
- Delete one of the node
```
$ minikube node delete minikube-m03
```
- The SeccompProfiles still have `minikube-m03` as finalizers
```
$ kubectl get seccompprofiles -n security-profiles-operator log-enricher-trace -o json | jq .metadata.finalizers
[
            "minikube-m03-delete",
            "minikube-m02-delete",
            "minikube-delete"
]
```
- I can't delete the seccompprofiles because the `"minikube-m03-delete"` is still exist (the old node that already deleted), unless if i patch it the finalizers like:
```
$ kubectl patch seccompprofiles log-enricher-trace -n security-profiles-operator -p '{"metadata":{"finalizers":[]}}' --type=merge
```

#### Anything else we need to know?:

I observed this on our Amazon EKS cluster, but for this testing i tested it on minikube with the same behavior. On our Amazon EKS clusters we have patching cycle that will delete the old node and bring up the new node (with new latest image). So after the patching the finalizers on the seccompprofiles still showing the old nodes, this make we can't delete the seccompprofles.

#### Environment:

- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Others:

### PR DESCRIPTION
**Pull request by Vincent056**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:

This pr fixes seccompprofiles deletion when a node is removed, we added a check to see if the node finalizer is a deleted node, if so, we remove such finalizer so the seccompprofile can be deleted without any issues.

The nodestatus controller detect the extra nodestatus object, and if it can't find the node associated with it, it will try to remove the node status object and node finalizer from corresponding the seccomp profile

#### Which issue(s) this PR fixes:

Fixes #1152

#### Does this PR have test?

Yes

#### Does this PR introduce a user-facing change?
NONE

```release-note
This pr fixes seccompprofiles deletion when a node is removed, we added a check to see if the node finalizer is a deleted node, if so, we remove such finalizer so the seccompprofile can be deleted without any issues.
```

### CODE REVIEW

File: vendor/sigs.k8s.io/controller-runtime/pkg/controller/controllerutil/controllerutil.go [hunk 1]
```diff
@@ -387,6 +390,57 @@ func ContainsFinalizer(o client.Object, finalizer string) bool {
 	return false
 }
 
+// RemoveDeletedNodeFinalizer removes finalizers from the object for deleted nodes.
+func RemoveDeletedNodeFinalizer(c client.Client, o client.Object) (error, string) {
```
Comment (jhrozek) [inline vendor/sigs.k8s.io/controller-runtime/pkg/controller/controllerutil/controllerutil.go:394]: Please don't patch files under `vendor/`, that should be an exception (e.g. if we need to bring in a patch that hasn't been released or merged yet)

File: internal/pkg/daemon/seccompprofile/seccompprofile.go [hunk 2]
```diff
@@ -468,6 +468,23 @@ func (r *Reconciler) reconcileDeletion(
 		return ctrl.Result{}, fmt.Errorf("deleting node status/finalizer for deleted SeccompProfile: %w", err)
 	}
 
+	// Remove the finalizer for deleted Nodes
+	err, nodeName := controllerutil.RemoveDeletedNodeFinalizer(r.client, sp)
```
[the remark this answers is not in the document]
Comment (Vincent056) [reply internal/pkg/daemon/seccompprofile/seccompprofile.go:472]: that's a good point, I will do that

File: internal/pkg/nodestatus/nodestatus.go [hunk 3]
```diff
@@ -128,6 +127,17 @@ func (nsf *StatusClient) statusObj(
 	}
 }
 
+func (nsf *StatusClient) statusObjNode(
+	nodeName string,
+) *secprofnodestatusv1alpha1.SecurityProfileNodeStatus {
+	return &secprofnodestatusv1alpha1.SecurityProfileNodeStatus{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      nsf.pol.GetName() + "-" + nodeName,
```
Comment (jhrozek) [inline internal/pkg/nodestatus/nodestatus.go:135]: could we reuse this in `statusObj` or just use `statusObj` instead?

File: internal/pkg/nodestatus/nodestatus.go [hunk 4]
```diff
@@ -165,6 +175,37 @@ func (nsf *StatusClient) Remove(ctx context.Context, c client.Client) error {
...
+		return fmt.Errorf("cannot clean up deleted node for %s: %w", nsf.pol.GetName(), err)
+	}
+
+	return nil
+}
+
+// remove the finalizer and delete the nodestatus for deleted node.
+func (nsf *StatusClient) deletedNodeCleanUp(ctx context.Context, c client.Client) error {
+	return util.Retry(func() error {
+		nodeName, err := util.RemoveDeletedNodeFinalizer(ctx, c, nsf.pol)
```
Comment (jhrozek) [inline internal/pkg/nodestatus/nodestatus.go:189]: should this be modified to handle more than one deleted node?

**Review by jhrozek — COMMENTED** [review]
nice progress, have you had a chance to test the PR manually yet while you're working on the tests?

File: internal/pkg/nodestatus/nodestatus.go (hunk 3 above)
Comment (Vincent056) [reply internal/pkg/nodestatus/nodestatus.go]: > nice progress, have you had a chance to test the PR manually yet while you're working on the tests?

I changed the implementation, I tested the patch manually, and it works

**Comment by Vincent056** [comment]
@jhrozek for e2e testing, since we can't delete node, what I did is to add a nodestatus object, and a finalizer to the seccomp profile to mimic a deleted node. The nodestatus controller should detect the extra nodestatus object, and if it can't find the node associated with it, it will try to remove the node status object and node finalizer from the seccomp profile

**Comment by saschagrunert** [comment]
> @saschagrunert Could you add me to the white list, so I don't have to bug someone every time for /ok-to-test

If you want to then we could add you as a reviewer to this project. This would also mean that you get notifications for other PRs as well :)

**Comment by jhrozek** [comment]
So the good thing is that the patch works. I tested on OpenShift, by scaling a `MachineSet` up, made sure both the finalizers and the `securityprofilenodestatus` objects reflect the new nodes and then scaled the `MachineSet` back down and made sure the finalizer and the status are gone.

To fix the CI, you should rebase on the current origin/main to fix some of the jobs and make sure the branch you rebase atop contains both PR #1253 and #1252. This should fix the `build/ubi-image`, `build/operator-image` and `test/e2e-flatcar` tests respectively.

Please also run `make bundle` and commit the diff to fix the `olm/tests` job.

**Comment by jhrozek** [comment]
also to get around the `release-note-needed` label, either please move the `NONE` text into the three-backticks below the `NONE` or even better, write a release note sentence into the backticks :-) (This is similar to how we keep a changelog in CO)

File: deploy/webhook-operator.yaml [hunk 5]
```diff
@@ -1202,7 +1202,15 @@ rules:
...
   - ""
   resources:
-  - services
+  - nodes
+  verbs:
+  - get
+  - list
+  - watch
+- apiGroups:
+  - admissionregistration.k8s.io
```
Comment (jhrozek) [inline deploy/webhook-operator.yaml:1211]: is this a merge conflict resolution? It doesn't seem like we need to modify the webhook-operator yaml to touch the `mutatingwebhookconfigurations` APIs do we?

File: internal/pkg/manager/nodestatus/nodestatus.go [hunk 6]
```diff
@@ -167,12 +169,28 @@ func (r *StatusReconciler) Reconcile(_ context.Context, req reconcile.Request) (
...
 	hasStatuses := len(nodeStatusList.Items)
 	wantsStatuses := spodDS.Status.DesiredNumberScheduled
-	if wantsStatuses != int32(hasStatuses) {
+	if wantsStatuses > int32(hasStatuses) {
 		logger.Info("Not updating policy: not all statuses are ready",
 			"has", hasStatuses, "wants", wantsStatuses)
 		// Don't reconcile again, let's just wait for another update
 		return reconcile.Result{}, nil
 	}
+	if wantsStatuses < int32(hasStatuses) {
```
Comment (jhrozek) [inline internal/pkg/manager/nodestatus/nodestatus.go:178]: I think it would be cleaner to do an else-if instead of two ifs
```
if wantsStatuses > int32(hasStatuses) {
 ...
} else if wantsStatuses < int32(hasStatuses) {
 ...
} 

// we know they are equal here
```
this is just to make it clear in the code that we are still continuing the same if

File: internal/pkg/manager/nodestatus/nodestatus.go [hunk 7]
```diff
@@ -167,12 +169,28 @@ func (r *StatusReconciler) Reconcile(_ context.Context, req reconcile.Request) (
...
+			return reconcile.Result{}, fmt.Errorf("cannot remove extra statuses: %w", err)
+		}
+		if nodeName != "" {
+			// remove the deleted node finalizer string from the profile
+			logger.Info("Removing finalizer from profile", "profile", prof.GetName(), "node", nodeName)
+			if err := util.RemoveFinalizer(ctx, r.client, prof, util.GetFinalizerNodeString(nodeName)); err != nil {
+				return reconcile.Result{}, fmt.Errorf("cannot remove finalizer from profile: %w", err)
+			}
+		}
+		return reconcile.Result{}, nil
```
Comment (jhrozek) [inline internal/pkg/manager/nodestatus/nodestatus.go:192]: since this code would only work for one node removal, should we force another reconcile loop here by returning:
```
return reconcile.Result{Requeue: true}, nil
```
this would allow us to both keep the logic simple to just remove one finalizer and one node object for 99% case where just one node goes away, but it would also cover the case where more nodes go away through multiple reconcile attempts.

**Review by jhrozek — COMMENTED** [review]
just some nits, but this works, great!

**Comment by Vincent056** [comment]
> > @saschagrunert Could you add me to the white list, so I don't have to bug someone every time for /ok-to-test
> 
> If you want to then we could add you as a reviewer to this project. This would also mean that you get notifications for other PRs as well :)

Thanks for adding me as a reviewer!

File: deploy/webhook-operator.yaml (hunk 5 above)
Comment (Vincent056) [reply deploy/webhook-operator.yaml:2099]: We didn't touch `mutatingwebhookconfigurations` in this PR, it seems to be auto-generated from `make bundle`

Comment (jhrozek) [reply deploy/webhook-operator.yaml:2099]: Ah I know now. Because adding the `node` RBAC rules added a first item to the `ClusterRole` the manager is running as, we need to adjust the webhook operator's kustomize manifests:
```
diff --git a/deploy/overlays/webhook/role.yaml b/deploy/overlays/webhook/role.yaml
index 97221506..f5c1444a 100644
--- a/deploy/overlays/webhook/role.yaml
+++ b/deploy/overlays/webhook/role.yaml
@@ -1,3 +1,3 @@
 ---
-- path: /rules/1
+- path: /rules/2
   op: remove
```
apply the above, then run `make deployments`. That did the trick for me.

File: .github/workflows/build.yml [hunk 8]
```diff
@@ -64,7 +64,7 @@ jobs:
       - uses: actions/checkout@v3
       - name: Set up Docker Buildx
         id: buildx
-        uses: docker/setup-buildx-action@c74574e6c82eeedc46366be1b0d287eff9085eb6
+        uses: docker/setup-buildx-action@8c0edbc76e98fa90f69d9a2c020dcb50019dc325
```
Comment (jhrozek) [inline .github/workflows/build.yml:67]: these don't belong in this PR (and neither does the commit message :-)). `git rebase origin/main` should get rid of these.

**Review by jhrozek — CHANGES_REQUESTED** [review]
the last issue is the namespaced webhook manifests and the commit message, otherwise I think we're good.
We're getting there, nice work!

**Comment by jhrozek** [comment]
btw I made a pass on the latest code functionality-wise and everything still works. So let's fix the nits and merge the patch!

**Comment by Vincent056** [comment]
> the last issue is the namespaced webhook manifests and the commit message, otherwise I think we're good. We're getting there, nice work!


Thanks for the help, I was able to fix the RBAC issue.
